### ISSUE TITLE
conditional type doesn't behave as expected on any

### ISSUE THREAD

**Issue by ori-sh** [microsoft/TypeScript#42805]
# Bug Report

### 🕗 Version & Regression Information

I encounter this when playing with TypeScript 4.1.3
And also tested on the  'Nightly' version in the playground.

I also reviewed the FAQ for entries about `any` type and conditional types and couldn't find a reference to something that clarifying this behavior.

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=4.3.0-dev.20210215#code/FAUwHgDg9gTgLgAjgTwiBA5ArgGxwSwGcALBAXgQDtccEAfBLSgExADN9KRmBuUSWIhRpMUONjxFiAHgAqAPnIJZCcHBAtCmGlIQB+KiABuIGAgBcyvsADGUSoUQBDSxjESCJaU8rJFFajweBAB6ENVIEBt1ZiQoBEIQdCckVHRTGFgEACMsREoxYk4AcwRiJwg0BwA6Wtt7RxzXdx0vRxgS+m1JEn8qGmCwhGKQITTVGEyzJy1wNGjuS1lxgHJAnBWEIioxBBnCfGLKJ2ycdDh44XQV9pKV6oAKACYAZiengEogA)

### 💻 Code

I'm trying to define a type that will narrow any given type to its not "Nullish" (`null` or `undefined`) version.
This example shows it works for `string | Nullish` for example and narrow it to `string` as expected but for some reason, it fails when it comes to `any` type.

```ts
export type Nullish = null | undefined;
export type NotNullish<T> = T extends Nullish ? never : T;

const a: NotNullish<any> = null; // expected to see a type error but nothing happens...
const b: NotNullish<string | Nullish> = null; // get type error as expected: Type 'null' is not assignable to type 'string'.(2322)
```

### 🙁 Actual behavior

Variable `a` can be assigned "Nullish" values without a type error.

### 🙂 Expected behavior

Expected that type for `a` won't allow assigning "Nullish" values and will show proper type error (same as appearing for `b` variable example).

**Comment by IllusionMH** [comment]
When you use any - you basically opt-out from type checking so `NotNullish<any>`  is still same as `any` so no type checking here and errors in case of `a`.

I guess that #29317 will allow `const a: NotNullish<unknown> = null;` to be an error, but it's unlikely that there can be error at all for `any`.

**Comment by RyanCavanaugh** [comment]
@IllusionMH is correct, you'd need negated types to represent this at all, and `any` being infectious through conditional types is the intended behavior.
